File: advent2017/extras/cxx/src/days/day17.cpp
// Architecture Notes:
// - This translation unit is documented for codebase-wide recoverability.
// - Solver stages are explicit to simplify verification and benchmarking.
// - Low-level optimizations are annotated with rationale and invariants.

#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "03a3d955b8799a90f1ff5a39479fde8e618f8ca3282d5b187186f2cf361abd32";
const std::string kExpectedPart1 = "808";
const std::string kExpectedPart2 = "47465686";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day17/d17_input.txt", "Day17/d17_input.txt", "../Day17/d17_input.txt", "../../Day17/d17_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

long long solve_part1(int step) {
    std::vector<int> buf{0};
    int pos = 0;
    for (int v = 1; v <= 2017; ++v) {
        pos = (pos + step) % static_cast<int>(buf.size()) + 1;
        buf.insert(buf.begin() + pos, v);
    }
    return buf[(pos + 1) % static_cast<int>(buf.size())];
}

long long solve_part2(int step) {
    int pos = 0;
    int val_after_zero = 0;
    int size = 1;
    for (int v = 1; v <= 50000000; ++v) {
        pos = (pos + step) % size + 1;
        if (pos == 1) val_after_zero = v;
        size++;
    }
    return val_after_zero;
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");
    const int step = std::stoi(raw);

    auto t0 = std::chrono::steady_clock::now();
    const std::string ans = (part == 1) ? std::to_string(solve_part1(step)) : std::to_string(solve_part2(step));
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=17 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day17.rs
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "03a3d955b8799a90f1ff5a39479fde8e618f8ca3282d5b187186f2cf361abd32";
const EXPECTED_PART1: &str = "808";
const EXPECTED_PART2: &str = "47465686";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day17/d17_input.txt", "Day17/d17_input.txt", "../Day17/d17_input.txt", "../../Day17/d17_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn solve_part1(step: usize) -> i64 {
    let mut buf: Vec<i64> = vec![0];
    let mut pos: usize = 0;
    for v in 1..=2017 {
        pos = (pos + step) % buf.len() + 1;
        buf.insert(pos, v as i64);
    }
    buf[(pos + 1) % buf.len()]
}

fn solve_part2(step: usize) -> i64 {
    let mut pos: usize = 0;
    let mut val_after_zero: i64 = 0;
    let mut size: usize = 1;
    for v in 1..=50_000_000usize {
        pos = (pos + step) % size + 1;
        if pos == 1 { val_after_zero = v as i64; }
        size += 1;
    }
    val_after_zero
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let step: usize = fs::read_to_string(&in_path).expect("read_to_string").trim().parse().expect("int");

    let t0 = Instant::now();
    let ans = if part == 1 { solve_part1(step).to_string() } else { solve_part2(step).to_string() };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=17 part={} runtime_ms={:.3}", part, ms);
}
